src/algorithms/lbfgs.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <RcppArmadillo.h>
+#include "../families/family.h"
+#include "wolfe_line_search.h"
+
+using namespace Rcpp;
+using namespace arma;
+
+// // LBFGS Algorithm.
+// // Solves argmin_z { f(z) + rho/2*||z-u||^2 }
+template <typename T>
+mat Family::lbfgs(const T& x, const mat&y, const double rho, const mat& u){
+
+  mat z(u);
+  mat g(u);
+  
+  const int max_iter = 2000;
+  const double tolerance = 1e-5;
+  const int l = 20;
+
+  std::deque <mat> dz;
```

Comment:
Why are you using `deque` instead of `vector`? AFAIK the only reason to use `deque` is when you want to insert things at the front of the vector, which I don't think we need?

Reply:
Since we only keep `l` elements to calculate hessian approximation, once we insert new values, there is no need for older valules. `deque` enables us to easily discard older values.

Reply:
On a second thought, I think `std::queue` would be even more suitable.

Reply:
Okay, just know that these containers sometimes mean worse performance. If you are only storing one value, maybe you could just use `old_dz` and `dz`.

Reply:
We need around 20-40 values so containers are inevitable.

Reply:
We need to access all the elements stored, which can't be done in queues. If use vectors, we have to keep track where the old elements are manually. Deque is the way.